updater/aws_test.go

```diff
@@ -15,6 +15,255 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestFilterAvailableUpdates(t *testing.T) {
+	checkPattern := "{\"update_state\": \"%s\", \"active_partition\": { \"image\": { \"version\": \"%s\"}}}"
+	instances := []instance{
+		{
+			instanceID:          "inst-id-1",
+			containerInstanceID: "cont-inst-1",
+		},
+		{
+			instanceID:          "inst-id-2",
+			containerInstanceID: "cont-inst-2",
+		},
+	}
+	cases := []struct {
+		name          string
+		invocationOut *ssm.GetCommandInvocationOutput
+		expectedOut   []instance
+	}{
+		{
+			name: "Update Available",
+			invocationOut: &ssm.GetCommandInvocationOutput{
+				Status:                aws.String("Success"),
+				StandardOutputContent: aws.String(fmt.Sprintf(checkPattern, updateStateAvailable, "v1.0.5")),
+			},
+			expectedOut: []instance{
+				{
+					instanceID:          "inst-id-1",
+					containerInstanceID: "cont-inst-1",
+					bottlerocketVersion: "v1.0.5",
+				},
+				{
+					instanceID:          "inst-id-2",
+					containerInstanceID: "cont-inst-2",
+					bottlerocketVersion: "v1.0.5",
+				},
+			},
+		},
+		{
+			name: "Update state Idle",
+			invocationOut: &ssm.GetCommandInvocationOutput{
+				Status:                aws.String("Success"),
+				StandardOutputContent: aws.String(fmt.Sprintf(checkPattern, updateStateIdle, "v1.1.1")),
+			},
+			expectedOut: []instance{},
+		},
+		{
+			name: "Update state Staged",
+			invocationOut: &ssm.GetCommandInvocationOutput{
+				Status:               aws.String("Success"),
+				StandardErrorContent: aws.String(fmt.Sprintf(checkPattern, updateStateStaged, "v1.1.1")),
+			},
+			expectedOut: []instance{},
+		},
+		{
+			name: "Update state Ready",
+			invocationOut: &ssm.GetCommandInvocationOutput{
+				Status:                aws.String("Success"),
+				StandardOutputContent: aws.String(fmt.Sprintf(checkPattern, updateStateReady, "v1.0.5")),
+			},
+			expectedOut: []instance{
+				{
+					instanceID:          "inst-id-1",
+					containerInstanceID: "cont-inst-1",
+					bottlerocketVersion: "v1.0.5",
+				},
+				{
+					instanceID:          "inst-id-2",
+					containerInstanceID: "cont-inst-2",
+					bottlerocketVersion: "v1.0.5",
+				},
+			},
+		},
```

Comment:
There is one more case missing where list has instances in different `UpdateStates`  and output contains only few instances from the list. Current test expectation can be easily met by a faulty code which only checks the first instance and returns all or none of the instances in output. Instead of iterating through all of them and filtering. 
You can either add to existing case or just add a single success case which considers different `UpdateStates`.  A case which considers a list of instances in all the different `UpdateStates` and output contains only instances in `Available` or `Ready` state.

Reply:
I replaced the `TestFilterAvailableUpdates` case with one that has a mixed set of instances. Keeping this and the previous version seemed redundant. However, happy to toss that back in if we want to validate how it handles each update state on an individual basis.